File: src/tests/prog.c
#include <stdio.h>
#include "reader.h"

#define STYLE_BOLD    "\033[1m"
#define STYLE_NO_BOLD "\033[22m"

int
main (int argc, char *argv[])
{
    // filename argument
    char *filename;
    if (argc == 2) {
        filename = argv [1];
    } else {
        fprintf (stderr, "missing zim file argument, see README.md\n");
        return 1;
    }


    // init reader
    GError* err = NULL;
    KiwixReader *reader = kiwix_reader_new (filename, &err);
    if (err != NULL)
    {
        fprintf (stderr, "%s\n", err->message);
        g_error_free (err);
        return 1;
    }


    // infos
    printf ("==INFOS==\n");

    const gchar *title = kiwix_reader_get_title (reader);
    const gchar *date = kiwix_reader_get_date (reader);
    const gchar *language = kiwix_reader_get_language (reader);
    int file_size = kiwix_reader_get_file_size (reader);
    const gchar *tags = kiwix_reader_get_tags (reader);
    const gchar *id = kiwix_reader_get_id (reader);
    const gchar *name = kiwix_reader_get_name (reader);
    const gchar *description = kiwix_reader_get_description (reader);
    const gchar *random_page_url = kiwix_reader_get_random_page_url (reader);
    unsigned int article_count = kiwix_reader_get_article_count (reader);
    unsigned int media_count = kiwix_reader_get_media_count (reader);
    unsigned int global_count = kiwix_reader_get_global_count (reader);
    const gchar *creator = kiwix_reader_get_creator (reader);
    const gchar *publisher = kiwix_reader_get_publisher (reader);

    //g_assert_cmpstr (title, ==, "Wikipédia");

    printf (STYLE_BOLD "Archive path:    " STYLE_NO_BOLD "%s\n", filename);
    printf (STYLE_BOLD "Title:           " STYLE_NO_BOLD "%s\n", title);
    printf (STYLE_BOLD "Date:            " STYLE_NO_BOLD "%s\n", date);
    printf (STYLE_BOLD "Language:        " STYLE_NO_BOLD "%s\n", language);
    printf (STYLE_BOLD "File size:       " STYLE_NO_BOLD "%d\n", file_size);
    printf (STYLE_BOLD "Tags:            " STYLE_NO_BOLD "%s\n", tags);
    printf (STYLE_BOLD "ID:              " STYLE_NO_BOLD "%s\n", id);
    printf (STYLE_BOLD "Name:            " STYLE_NO_BOLD "%s\n", name);
    printf (STYLE_BOLD "Description:     " STYLE_NO_BOLD "%s\n", description);
    printf (STYLE_BOLD "Random page url: " STYLE_NO_BOLD "%s\n", random_page_url);
    printf (STYLE_BOLD "Article count:   " STYLE_NO_BOLD "%u\n", article_count);
    printf (STYLE_BOLD "Media count:     " STYLE_NO_BOLD "%u\n", media_count);
    printf (STYLE_BOLD "Global count:    " STYLE_NO_BOLD "%u\n", global_count);
    printf (STYLE_BOLD "Creator:         " STYLE_NO_BOLD "%s\n", creator);
    printf (STYLE_BOLD "Publisher:       " STYLE_NO_BOLD "%s\n", publisher);


    // suggestions
    printf ("==SUGGESTIONS==\n");

    kiwix_reader_search_suggestions_smart (reader, "ab", 20);

    char *suggestion;
    char *suggestion_url;
    while (kiwix_reader_get_next_suggestion (reader, &suggestion, &suggestion_url))
    {
        printf (STYLE_BOLD "Suggestion: " STYLE_NO_BOLD "%s, ", suggestion);
        printf (STYLE_BOLD "Suggestion url: " STYLE_NO_BOLD "%s\n", suggestion_url);
        g_free (suggestion);
        g_free (suggestion_url);
    }


    // favicon
    printf ("==FAVICON==\n");

    gsize size;
    char* data = NULL;
    char* mime_type = NULL;
    kiwix_reader_get_faveicon (reader, &data, &size, &mime_type);

    printf (STYLE_BOLD "Faveicon size     : " STYLE_NO_BOLD "%lu\n", size);
    printf (STYLE_BOLD "Faveicon mime type: " STYLE_NO_BOLD "%s\n", mime_type);

    // write faveicon to file
    FILE *f;
    f = fopen("faveicon-from-c.png", "wb");
    fwrite (data, size, 1, f);
    fclose (f);

    return 0;
}

File: src/tests/prog.py
#!/usr/bin/env python3

import sys
import gi
gi.require_version('Kiwix', '1.0')
from gi.repository import Kiwix
from gi.repository import GLib

STYLE_BOLD    = "\033[1m"
STYLE_NO_BOLD = "\033[22m"

def main (argc, argv):
    # filename argument
    if argc == 2:
        filename = argv[1]
    else:
        print ("missing zim file argument, see README.md", file=sys.stderr)
        exit (1)


    # init reader
    try:
        reader = Kiwix.Reader.new (filename)
    except GLib.Error as e:
        print (e, file=sys.stderr)
        exit(1)


    # infos
    print ("==INFOS==");

    title = reader.get_title ()
    date = reader.get_date ()
    language = reader.get_language ()
    file_size = reader.get_file_size ()
    tags = reader.get_tags ()
    id = reader.get_id ()
    name = reader.get_name ()
    description = reader.get_description ()
    random_page_url = reader.get_random_page_url ()
    article_count = reader.get_article_count ()
    media_count = reader.get_media_count ()
    global_count = reader.get_global_count ()
    creator = reader.get_creator ()
    publisher = reader.get_publisher ()

    #assert( title == "Wikipédia" )

    print (STYLE_BOLD + "Archive path:    " + STYLE_NO_BOLD + filename);
    print (STYLE_BOLD + "Title:           " + STYLE_NO_BOLD + title);
    print (STYLE_BOLD + "Date:            " + STYLE_NO_BOLD + date);
    print (STYLE_BOLD + "Language:        " + STYLE_NO_BOLD + language);
    print (STYLE_BOLD + "File size:       " + STYLE_NO_BOLD + str(file_size));
    print (STYLE_BOLD + "Tags:            " + STYLE_NO_BOLD + tags);
    print (STYLE_BOLD + "ID:              " + STYLE_NO_BOLD + id);
    print (STYLE_BOLD + "Name:            " + STYLE_NO_BOLD + name);
    print (STYLE_BOLD + "Description:     " + STYLE_NO_BOLD + description);
    print (STYLE_BOLD + "Random page url: " + STYLE_NO_BOLD + random_page_url);
    print (STYLE_BOLD + "Article count:   " + STYLE_NO_BOLD + str(article_count));
    print (STYLE_BOLD + "Media count:     " + STYLE_NO_BOLD + str(media_count));
    print (STYLE_BOLD + "Global count:    " + STYLE_NO_BOLD + str(global_count));
    print (STYLE_BOLD + "Creator:         " + STYLE_NO_BOLD + creator);
    print (STYLE_BOLD + "Publisher:       " + STYLE_NO_BOLD + publisher);


    # suggestions
    print ("==SUGGESTIONS==");

    reader.search_suggestions_smart ("ab", 20)

    next = True
    while next:
        next, suggestion, suggestion_url = reader.get_next_suggestion ()
        if next:
            print (STYLE_BOLD + "Suggestion: " + STYLE_NO_BOLD + suggestion + ", ", end='')
            print (STYLE_BOLD + "Suggestion url: " + STYLE_NO_BOLD + suggestion_url)


    # favicon
    print ("==FAVICON==")

    result, data, mime_type = reader.get_faveicon ()
    size = len(data)

    print (STYLE_BOLD + "Favicon size:      " + STYLE_NO_BOLD + str(size))
    print (STYLE_BOLD + "Favicon mime type: " + STYLE_NO_BOLD + mime_type)

    # write faveicon to file
    f = open ('faveicon-from-python.png', 'wb')
    f.write (data)
    f.close ()


if __name__ == "__main__":
    main (len(sys.argv), sys.argv)
